src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+
+If instead the 4th thread's store happened before the 5th thread's load, we'd have gotten the correct answer 5.
+
+
+This is a *data race* bugs. They are _very_ difficult to detect, because they depend on the scheduling of the threads.
+
+Luckily, we've figured out how to avoid these problems, with some "concurrency ground rules":
+
+ 1. Multiple threads can read the same data, if nobody can modify it.
+ 1. If a thread can read data that another thread can modify, the data must be wrapped in a mutex. [# More on mutexes below.]
+ 1. If data is only visible to one thread, it can access that data freely.
+
+
+We usually need proper fear, respect, and discipline to stick to these rules. But when a language enforces these rules, we have *fearless concurrency*.
+
+
+<slice/>
+
+
+# Fearless Concurrency
+
+Plenty of languages (Clojure, Pony, Erlang, Rust, Dart, some others [# See Renato Athaydes' [Fearless concurrency: how Clojure, Rust, Pony, Erlang and Dart let you achieve that.](https://sites.google.com/a/athaydes.com/renato-athaydes/posts/fearlessconcurrencyhowclojurerustponyerlanganddartletyouachievethat)]) offer fearless concurrency, but we'll use Pony and Rust as examples.
+
+
+## Message Passing with Pony
+
+Pony doesn't let us access data from the outside scope, like in our above C examples. Instead, we "send" data between threads.
+
+
+We can send data if *either* of these is true:
+
+ * There are no other references to the data (it has the `iso` permission). [# A "permission" is something that travels along with a pointer at compile time, which affects how you may use that pointer. For example, Rust's `&` vs `&mut`, C++'s `*` vs `const *`, Vale's `&` vs `*`.]
+ * It is deeply immutable (it has the `imm` permission).
+
+To send an object to another thread, it must have the `imm` or `iso` permission.
+
+
+It's a restrictive system, but data races are impossible. Huzzah!
+
+
+<slice/>
+
+
+## Why no Mutex?
+
+Having a mutex is not enough to prevent data races.
+
+
+For example, in Java, we would use a mutex like this:
+
+```java
+synchronized (myObj) {
+  Counter counter = myObj.myCounter;
+  myCounter.number = myCounter.number + 1;
+}
+```
+
+This code looks good; the `synchronized` block will lock the mutex in `myObj` [# In Java, every object is (or contains) a mutex.] and then unlock it when the block ends.
+
+
+However, the programmer can still mess this up: they can access `counter` after the block!
+
+```java
+Counter counter;
+synchronized (myObj) {
+  counter = myObj.myCounter;
+  myCounter.number = myCounter.number + 1;
+}
+myCounter.number = myCounter.number + 7;
+```
+
+
+The problem is that references can *outlive* the mutex lock.
+
+
+Luckily, there's a way to prevent that. Enter Rust!
+
+
+<slice/>
+
+
+## Mutexes with Rust
+
+Just like how Pony tracks permissions, Rust tracks lifetimes, and can ensure that no reference escapes a mutex lock.
```

Comment:
Unlike, the're pretty different systems imo

Reply:
Removed